kube-runtime/src/events.rs

```diff
@@ -289,30 +426,106 @@ mod test {
             "Sending kubernetes to detention without namespace"
         );
 
+        recorder
+            .publish(
+                &Event {
+                    type_: EventType::Normal,
+                    reason: "VeryCoolServiceNoNamespace".into(),
+                    note: Some("Sending kubernetes to detention without namespace twice".into()),
+                    action: "Test event - plz ignore".into(),
+                    secondary: None,
+                },
+                &s.object_ref(&()),
+            )
+            .await?;
+
+        let event_list = events.list(&Default::default()).await?;
+        let found_event = event_list
+            .into_iter()
+            .find(|e| std::matches!(e.reason.as_deref(), Some("VeryCoolServiceNoNamespace")))
+            .unwrap();
+        assert!(found_event.series.is_some());
+        Ok(())
+    }
+
+    #[tokio::test]
+    #[ignore = "needs cluster (creates an event for the default kubernetes service)"]
+    async fn event_recorder_cache_retain() -> Result<(), Box<dyn std::error::Error>> {
+        let client = Client::try_default().await?;
+
+        let svcs: Api<Service> = Api::namespaced(client.clone(), "default");
+        let s = svcs.get("kubernetes").await?; // always a kubernetes service in default
+
+        let reference = s.object_ref(&());
+        let reporter: Reporter = "kube".into();
+        let ev = Event {
+            type_: EventType::Normal,
+            reason: "TestCacheTtl".into(),
+            note: Some("Sending kubernetes to detention".into()),
+            action: "Test event - plz ignore".into(),
+            secondary: None,
+        };
+        let key = EventKey {
+            event_type: ev.type_,
+            action: ev.action.clone(),
+            reason: ev.reason.clone(),
+            reporting_controller: reporter.controller.clone(),
+            regarding: Reference(reference.clone()),
+            reporting_instance: None,
+            related: None,
+        };
+
+        let reporter = Reporter {
+            controller: "kube".into(),
+            instance: None,
+        };
+        let recorder = Recorder::new(client.clone(), reporter);
+
+        recorder.publish(&ev, &s.object_ref(&())).await?;
+        let now = Utc::now();
+        let past = now - Duration::minutes(10);
+        recorder.cache.write().await.entry(key).and_modify(|e| {
+            e.event_time = Some(MicroTime(past));
+        });
+
+        recorder.publish(&ev, &s.object_ref(&())).await?;
+
+        let events: Api<K8sEvent> = Api::namespaced(client, "default");
+        let event_list = events.list(&Default::default()).await?;
+        let found_event = event_list
+            .into_iter()
+            .find(|e| std::matches!(e.reason.as_deref(), Some("TestCacheTtl")))
+            .unwrap();
+        assert_eq!(found_event.note.unwrap(), "Sending kubernetes to detention");
+        assert!(found_event.series.is_none());
```

Comment:
minor q: shouldn't series be a `Some` in this case? we've published the same event twice?

Reply:
Here we are testing the retain cache policy. I modify the event to be removed by the TTL and test that a new one is created.